incubator/hnc/internal/validators/namespace.go

```diff
@@ -64,19 +64,15 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
 // handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
 // tested (ie without constructing a full admission.Request).
 func (v *Namespace) handle(req *nsRequest) admission.Response {
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
 	parent := req.ns.Annotations[api.SubnamespaceOf]
-	if parent != "" && req.op == v1beta1.Delete {
+	if parent != "" && v.Forest.Get(parent).HasAnchor(req.ns.Name) && req.op == v1beta1.Delete {
```

Comment:
This needs a lot more commenting IMO plus maybe a bit of refactoring. E.g. something like:

```
// If the anchor doesn't exist, we want to allow it to be deleted anyway because ...
// See issue #xxx
anchorExists := v.Forest.Get(parent).HasAnchor(req.ns.Name) <-- should be false if this isn't a subns, I think
if req.op == v1beta1.Delete && anchorExists {
   msg := ...
   return deny(forbidden, msg)
}
```

Reply:
sg. updated!